File: src/Functional/Elixir/example3.exs
defmodule ReportFactory do
  def generate(:pdf), do: PDFReport.generate()
  def generate(:html), do: HTMLReport.generate()
end

defmodule PDFReport do
  def generate() do
    IO.puts "Generating PDF report"
  end
end

defmodule HTMLReport do
  def generate() do
    IO.puts "Generating HTML report"
  end
end

# Usage
ReportFactory.generate(:pdf)
ReportFactory.generate(:html)

File: src/Functional/Haskell/Example3.hs
class Report r where
  generate :: r -> IO ()

data PDFReport = PDFReport
data HTMLReport = HTMLReport

instance Report PDFReport where
  generate _ = putStrLn "Generating PDF report"

instance Report HTMLReport where
  generate _ = putStrLn "Generating HTML report"

class ReportFactory rf where
  createReport :: rf -> IO ()

data PDFReportFactory = PDFReportFactory
data HTMLReportFactory = HTMLReportFactory

instance ReportFactory PDFReportFactory where
  createReport _ = do
    let report = PDFReport
    generate report

instance ReportFactory HTMLReportFactory where
  createReport _ = do
    let report = HTMLReport
    generate report

useFactory :: (ReportFactory rf) => rf -> IO ()
useFactory factory = createReport factory

main :: IO ()
main = do
  useFactory PDFReportFactory
  useFactory HTMLReportFactory
